devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,402 @@
+    if no_of_space_dims > 1:
+        args.extend(list(objs['grid'].distributor.topology)[::-1])
+
+    # Number of degrees of freedom per node
+    args.append(1)
+    # "Stencil width" -> size of overlap
+    args.append(dmda.stencil_width)
+    args.extend([Null for _ in range(no_of_space_dims)])
+
+    # The distributed array object
+    args.append(Byref(dmda))
+
+    # The PETSc call used to create the DMDA
+    dmda = petsc_call('DMDACreate%sd' % no_of_space_dims, args)
+
+    return dmda
+
+
+def build_solver_objs(target):
+    name = target.name
+
+    return {
+        'Jac': Mat(name='J_%s' % name),
+        'x_global': Vec(name='x_global_%s' % name),
+        'x_local': Vec(name='x_local_%s' % name, liveness='eager'),
+        'b_global': Vec(name='b_global_%s' % name),
+        'b_local': Vec(name='b_local_%s' % name, liveness='eager'),
+        'ksp': KSP(name='ksp_%s' % name),
+        'pc': PC(name='pc_%s' % name),
+        'snes': SNES(name='snes_%s' % name),
+        'X_global': Vec(name='X_global_%s' % name),
+        'Y_global': Vec(name='Y_global_%s' % name),
+        'X_local': Vec(name='X_local_%s' % name, liveness='eager'),
+        'Y_local': Vec(name='Y_local_%s' % name, liveness='eager')
+    }
+
+
+def generate_solver_calls(solver_objs, objs, matvec, target):
+
+    dmda = objs['da_so_%s' % target.space_order]
+
+    solver_params = matvec.expr.rhs.solver_parameters
+
+    snes_create = petsc_call('SNESCreate', [objs['comm'], Byref(solver_objs['snes'])])
+
+    snes_set_dm = petsc_call('SNESSetDM', [solver_objs['snes'], dmda])
+
+    create_matrix = petsc_call('DMCreateMatrix', [dmda, Byref(solver_objs['Jac'])])
+
+    # NOTE: Assumming all solves are linear for now.
+    snes_set_type = petsc_call('SNESSetType', [solver_objs['snes'], 'SNESKSPONLY'])
+
+    global_x = petsc_call('DMCreateGlobalVector',
+                          [dmda, Byref(solver_objs['x_global'])])
+
+    local_x = petsc_call('DMCreateLocalVector',
+                         [dmda, Byref(solver_objs['x_local'])])
+
+    global_b = petsc_call('DMCreateGlobalVector',
+                          [dmda, Byref(solver_objs['b_global'])])
+
+    local_b = petsc_call('DMCreateLocalVector',
+                         [dmda, Byref(solver_objs['b_local'])])
+
+    snes_get_ksp = petsc_call('SNESGetKSP',
+                              [solver_objs['snes'], Byref(solver_objs['ksp'])])
+
+    vec_replace_array = petsc_call(
+        'VecReplaceArray', [solver_objs['x_local'],
+                            FieldFromPointer(target._C_field_data, target._C_symbol)])
+
+    ksp_set_tols = petsc_call(
+        'KSPSetTolerances', [solver_objs['ksp'], solver_params['ksp_rtol'],
+                             solver_params['ksp_atol'], solver_params['ksp_divtol'],
+                             solver_params['ksp_max_it']])
+
+    ksp_set_type = petsc_call(
+        'KSPSetType', [solver_objs['ksp'], solver_mapper[solver_params['ksp_type']]])
```

Comment:
for homogeneity, the closing parenthesis should go on the next line when you place the arguments on a new line

that is this:

```
x = foo(
    arg0, arg1, ...
)
```

rather than

```
x = foo(
    arg0, arg1, ...)
```

Obviously, this is all just boring nitpicking